source/simulator/solver.cc

```diff
@@ -287,23 +262,88 @@ namespace aspect
         utmp += src.block(0);
       }
 
-      // now either solve with the top left block (if do_solve_A==true)
+      A_inv_operator.vmult(dst.block(0), utmp);
+    }
+
+
+    /**
+      * This class is used in the implementation of the right preconditioner
+      * as an approximation for the inverse of the velocity (A) block.
+      * This operator can either just apply the preconditioner (AMG)
+      * or perform an inner CG solve with the same preconditioner.
+      * Here, the Schur complement is approximated by
+      * the pressure mass matrix weighted by the inverse of viscosity and
+      * the inverse is computed with a CG solve preconditioned by
+      * PreconditionerMp passed to the constructor.
+      */
+    template <class PreconditionerA>
+    class InverseVelocityBlock
+    {
+      public:
+        /**
+           * @param do_solve_A A flag indicating whether we should actually solve with
+           *     the matrix $A$, or only apply one preconditioner step with it.
+           * @param A_block_is_symmetric A flag indicating whether the matrix $A$ is symmetric.
+           * @param A_block_tolerance The tolerance for the CG solver which computes
+           *     the inverse of the A block.
+        */
+        InverseVelocityBlock(const TrilinosWrappers::SparseMatrix &matrix,
+                             const PreconditionerA &preconditioner,
+                             const bool do_solve_A,
+                             const bool A_block_is_symmetric,
```

Comment:
If you do document parameters, document them all :-) Might also be nice to start with a one-sentence summary of what the function does (here, "Constructor.").